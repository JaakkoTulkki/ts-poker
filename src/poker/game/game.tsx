import { Card, Deck } from '../deck/deck';
import { Validator, PokerValidator } from './validator';

export class Hand {
    constructor(private cards: Card[]) {
    }

    getCards(): Card[] {
        return this.cards;
    }
}

export class Poker {
    private deck: Deck;
    private hand: Hand;
    private hasSwapped = false;

    constructor(private validator: Validator = new PokerValidator()) {
        this.deck = new Deck();
        this.deck.shuffle();
        this.makeInitialHand();
    }

    deal() {
        if (this.hasSwapped) {
            throw new Error('You can only change your cards once');
        }

        let newCards: Card[] = [];
        for (let card of this.hand.getCards()) {
            if (card.shouldFold()) {
                let newCard: Card|undefined = this.deck.getCards().shift();
                if (!newCard) {
                    throw new Error('You ran out of cards');
                }
                newCards.push(newCard);
            } else {
                newCards.push(card);
            }
        }
        this.hand = new Hand(newCards);
        this.hasSwapped = true;
    }

    getHand(): Hand {
        return this.hand;
    }

    evaluateHand(): string {
        return this.validator.getBest(this.hand);
    }

    private makeInitialHand() {
        this.hand = new Hand(this.deck.getCards().splice(0, 5));
    }
}