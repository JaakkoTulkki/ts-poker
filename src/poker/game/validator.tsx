import { Hand } from './game';
import { Card } from '../deck/deck';

export const HIGH_CARD = 'HIGH_CARD';
export const PAIR = 'PAIR';
export const TWO_PAIRS = 'TWO_PAIRS';
export const THREE_OF_KIND = 'THREE_OF_KIND';
export const STRAIGHT = 'STRAIGHT';
export const FLUSH = 'FLUSH';
export const FULL_HOUSE = 'FULL_HOUSE';
export const FOUR_OF_KIND = 'FOUR_OF_KIND';
export const ROYAL_FLUSH = 'ROYAL_FLUSH';

export interface Validator {
    getBest(hand: Hand): string;
}

export class PokerValidator implements Validator {
    getBest(hand: Hand): string {
        const cards = hand.getCards();
        const self = this;
        const order = [
            [this.isRoyalFlush.bind(self), ROYAL_FLUSH],
            [this.isFourOfKind.bind(self), FOUR_OF_KIND],
            [this.isFullHouse.bind(self), FULL_HOUSE],
            [this.isFlush.bind(self), FLUSH],
            [this.isStraight.bind(self), STRAIGHT],
            [this.isThreeOfKind.bind(self), THREE_OF_KIND],
            [this.isTwoPairs.bind(self), TWO_PAIRS],
            [this.isPair.bind(self), PAIR],
        ];
        for (let validationPair of order) {
            let [validator, returnValue] = validationPair;
            if (typeof validator === 'function' && typeof returnValue === 'string' && validator(cards)) {
                return returnValue;
            }
        }
        return HIGH_CARD;
    }

    private isPair(cards: Card[]): boolean {
        const values = this.getUniqueValues(cards);
        return values.size === 4;
    }

    private isTwoPairs(cards: Card[]): boolean {
        const values = cards.map((card) => card.value);
        const uniqueValues = this.countNumbers(values);
        let numberOfPairs = 0;
        Object.keys(uniqueValues).forEach(key => {
            if (uniqueValues[key] === 2) {
                numberOfPairs += 1;
            }
        });
        return numberOfPairs === 2;
    }

    private isThreeOfKind(cards: Card[]): boolean {
        const values = cards.map((card) => card.value);
        const uniqueValues = this.countNumbers(values);
        let found = false;
        Object.keys(uniqueValues).forEach(key => {
            if (uniqueValues[key] === 3) {
                found = true;
            }
        });
        return found;
    }

    private isFlush(cards: Card[]): boolean {
        const suites = this.getUniqueSuites(cards);
        return suites.size === 1;
    }

    private isStraight(cards: Card[]): boolean {
        const uniqueValues = Array.from(this.getUniqueValues(cards));
        if (uniqueValues.length !== 5) {
            return false;
        }
        let highestValue = Math.max(...uniqueValues);
        let lowestValue = Math.min(...uniqueValues);

        if ((highestValue - lowestValue) === 4) {
            return true;
        }
        if (lowestValue === 1) {
            const valuesWithoutAce = uniqueValues.filter(value => value !== 1);
            highestValue = Math.max(...valuesWithoutAce);
            lowestValue = Math.min(...valuesWithoutAce);
            if ((highestValue - lowestValue) === 3) {
                return true;
            }
        }
        return false;
    }

    private isFourOfKind(cards: Card[]) {
        const values = cards.map(card => card.value);
        if (new Set(values).size !== 2) {
            return false;
        }
        const valuesCounter = this.countNumbers(values);
        return Object.keys(valuesCounter).every(key => {
            const counter = valuesCounter[key];
            return counter === 1 || counter === 4;
        });
    }

    private isFullHouse(cards: Card[]): boolean {
        const values = cards.map(card => card.value);
        const valuesCounter = this.countNumbers(values);
        if (Object.keys(valuesCounter).length !== 2) {
            return false;
        }
        return Object.keys(valuesCounter).every(key => {
            const counter = valuesCounter[key];
            return counter === 2 || counter === 3;
        });
    }

    private countNumbers(values: number[]) {
        return values.reduce((acc, value) => {
            if (!acc[value]) {
                acc[value] = 1;
            } else {
                acc[value] += 1;
            }
            return acc;
        },                   {});
    }

    private isRoyalFlush(cards: Card[]) {
        return this.isFlush(cards) && this.isStraight(cards);
    }

    private getUniqueValues(cards: Card[]): Set<number> {
        const values = new Set();
        for (let card of cards) {
            values.add(card.value);
        }
        return values;
    }

    private getUniqueSuites(cards: Card[]): Set<string> {
        const suites = new Set();
        for (let card of cards) {
            suites.add(card.suite);
        }
        return suites;
    }
}