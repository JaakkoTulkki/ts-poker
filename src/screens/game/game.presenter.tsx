import * as React from 'react';
import { Hand, Poker } from '../../poker/game/game';
import { Card as CardDeck } from '../../poker/deck/deck';

interface CardProps {
    card: CardDeck;
}

interface CardState {
    swap: boolean;
}

interface CardValueProps {
    suite: string;
    value: number;
}

const cardValues = {
    1: 'A',
    11: 'J',
    12: 'Q',
    13: 'K',
};

export const CardValue: React.SFC<CardValueProps> = (props) => (
    <div className="card-value">
        <img src={`/Card_${props.suite}.svg`} />
        <div className="card-value--value">
            {cardValues[props.value] ? cardValues[props.value] : props.value}
        </div>
    </div>
);

export class Card extends React.Component<CardProps, CardState> {
    constructor(props: CardProps) {
        super(props);
        props.card.fold(); // card is folded by default
        this.state = {
            swap: props.card.shouldFold()
        };
        this.select = this.select.bind(this);
    }

    select() {
        this.setState((state, props) => {
            props.card.fold();
            return {
                swap: props.card.shouldFold()
            };
        });
    }

    render() {
        return (
            <div className={`card`}>
                <div
                    className={`card__elem ${!this.state.swap ? 'keep' : ''}`}
                    data-card-value={this.props.card.toString()}
                    data-selected={this.state.swap}
                    onClick={this.select}
                    data-test-id="card"
                >
                    <CardValue value={this.props.card.value} suite={this.props.card.suite} />
                </div>
                {`${!this.state.swap ? 'Keep' : 'Swap'}`}
            </div>);
    }
}

export class Game extends React.Component<{}, {hand: Hand, poker: Poker, hasDrawn: boolean}> {
    constructor(props: object) {
        super(props);
        const poker = new Poker();
        const hand: Hand = poker.getHand();
        this.state = {
            poker,
            hand,
            hasDrawn: false,
        };
        this.deal = this.deal.bind(this);
        this.restartGame = this.restartGame.bind(this);
    }

    restartGame() {
        const poker = new Poker();
        const hand: Hand = poker.getHand();
        this.setState(() => ({
            poker,
            hand,
            hasDrawn: false,
        }));
    }

    deal() {
        this.setState((state, props) => {
            state.poker.deal();
            return {
                hand: state.poker.getHand(),
                hasDrawn: true,
            };
        });
    }

    render() {
        return (
            <div>
                <div>
                    {this.state.hand.getCards().map((card: CardDeck) => (
                        <Card key={card.toString()} card={card}/>))}
                </div>
                {!this.state.hasDrawn &&
                    <button id="deal" onClick={this.deal}>Deal New Cards</button>
                }
                {this.state.hasDrawn &&
                    <div>
                        <div id="result">Your hand is {this.state.poker.evaluateHand()}</div>
                        <button id="new-cards" onClick={this.restartGame}>Start Again</button>
                    </div>
                }
            </div>
        );
    }
}
