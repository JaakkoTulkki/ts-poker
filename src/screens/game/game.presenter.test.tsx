import * as React from 'react';
import { mount } from 'enzyme';
import { Game, Card } from './game.presenter';

test('it should render five cards', () => {
    const expectedNumOfCards = 5;
    const wrapper = mount(<Game />);
    const actualLength = wrapper.find(Card).length;
    expect(actualLength).toEqual(expectedNumOfCards);
});

test('each card should be different', () => {
    const expectedNumOfValues = 5;
    const wrapper = mount(<Game />);
    const cards = new Set(wrapper.find('.card').map(component => component.props()['data-card-value']));
    expect(cards.size).toBe(expectedNumOfValues);
});

test('clicking cards should toggle them (select/unselect)', () => {
    const wrapper = mount(<Game />);
    let firstCard = wrapper.find('.card').at(1);
    let thirdCard = wrapper.find('.card').at(3);

    expect(firstCard.props()['data-selected']).toBe(true);
    expect(thirdCard.props()['data-selected']).toBe(true);

    firstCard.simulate('click');
    thirdCard.simulate('click');

    firstCard = wrapper.find('.card').at(1);
    thirdCard = wrapper.find('.card').at(3);
    expect(firstCard.props()['data-selected']).toBe(false);
    expect(thirdCard.props()['data-selected']).toBe(false);
});

test('clicking two cards and then swapping should keep the two selected cards', () => {
    const wrapper = mount(<Game />);
    const originalCardValues = wrapper.find('.card').map((component) => component.props()['data-card-value']);
    const firstCard = wrapper.find('.card').at(1);
    const thirdCard = wrapper.find('.card').at(3);
    const unSelectedCard = wrapper.find('.card').at(4);
    firstCard.simulate('click');
    thirdCard.simulate('click');

    wrapper.find('button #deal').simulate('click');

    const newCardValues = wrapper.find('.card').map((component) => component.props()['data-card-value']);
    expect(newCardValues).not.toEqual(originalCardValues);
    expect(newCardValues.indexOf(firstCard.props()['data-card-value']) > -1).toBeTruthy();
    expect(newCardValues.indexOf(thirdCard.props()['data-card-value']) > -1).toBeTruthy();
    expect(newCardValues.indexOf(unSelectedCard.props()['data-card-value']) > -1).toBe(false);
});

test('you should not be able to draw twice', () => {
    const wrapper = mount(<Game />);
    wrapper.find('button #deal').simulate('click');
    expect(wrapper.find('button #deal').length).toBe(0);
});

test('Selecting two cards and drawing should show the best score', () => {
    const wrapper = mount(<Game />);
    wrapper.find('.card').at(1).simulate('click');
    wrapper.find('.card').at(3).simulate('click');
    wrapper.find('button #deal').simulate('click');

    const infoText = 'Your hand is ';
    const actualText = wrapper.find('#result').text();
    expect(actualText.startsWith(infoText)).toBeTruthy();
    expect(actualText.length).toBeGreaterThan(infoText.length);
});
