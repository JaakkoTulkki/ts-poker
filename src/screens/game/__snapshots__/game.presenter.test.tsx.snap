// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardValue should display correct suite with correct value 1`] = `
<div
  className="card-value"
>
  <img
    src="/Card_heart.svg"
  />
  <div
    className="card-value--value"
  >
    5
  </div>
</div>
`;
